sklearn/neural_network/multilayer_perceptron.py

```diff
@@ -507,12 +509,14 @@ def _fit_stochastic(self, X, y, activations, deltas, coef_grads,
         try:
             for it in range(self.max_iter):
                 if self.shuffle:
-                    X, y = shuffle(X, y, random_state=self._random_state)
+                    idx = shuffle(idx, random_state=self._random_state)
                 accumulated_loss = 0.0
                 for batch_slice in gen_batches(n_samples, batch_size):
-                    activations[0] = X[batch_slice]
```

Comment:
Maybe add a fast path here,
```py
if self.shuffle
    X_batch = safe_indexing(X, idx[batch_slice])
    y_batch = y[idx[batch_slice]]
else:
    X_batch = X[batch_slice]
    y_batch = y[batch_slice]
```
to avoid using fancy indexing when it's not needed.

Reply:
Thanks for the proposal.

The numpy indexing of random ordered integer indices is slightly slower than indexing of a slice, so it is a good idea to add a fast-path for non-shuffled data (13.85s vs 12.13s in my benchmark)